### PR TITLE
Status subcommand reporting agent status for otel mode - Phase 2

### PR DESCRIPTION
**Pull request by michalpristas**
This is a phase 2 of making status for otel work properly

**Bolds** are differences with previous Phase

Phase 1:
- agent capable of running in normal OR otel mode 
- coordinator not running at all
- otel collector running in a separate thread
- otel agent not able to report state

**Phase 2: We are here** 
- agent capable of running in normal OR otel mode
- **coordinator running with dumb config provider (no config)**
- otel collector running in a separate thread
- **status reports only agent status (otel status not reported)**

Fixes partially: https://github.com/elastic/elastic-agent/issues/3872

### CODE REVIEW

**Comment by pierrehilbert** [comment]
Failure related to https://github.com/elastic/elastic-agent/issues/3342

**Comment by ycombinator** [comment]
@michalpristas Just to clarify, phase 1 was implemented as part of https://github.com/elastic/elastic-agent/pull/3800, right?

**Comment by ycombinator** [comment]
Also, I'm confused by this:

> Phase 1
> otel collector running in a separate thread

> Phase 2
> **otel collector running in a separate thread**

What is the difference (indicated by the bolding in Phase 2)?

**Comment by michalpristas** [comment]
my bad

**Comment by ycombinator** [comment]
I built Agent from this PR's branch, extracted it, and ran `$ ./elastic-agent -c otel.yml status`.  I'm getting the following output:

```
Error: failed to communicate with Elastic Agent daemon: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial unix /private/tmp/elastic-agent-8.12.0-SNAPSHOT-darwin-aarch64/elastic-agent.sock: connect: no such file or directory"
For help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.12/fleet-troubleshooting.html
```

Is this expected behavior, even with the changes in this PR?

_Please ignore the `8.12` references above; that's because we're currently having an issue with building Agent from `main` with `EXTERNAL=true`; the workaround is to also specify `BEAT_VERSION=8.12.0`._

---

BTW, I re-wrote the issue that this PR is supposed to resolve to include more details: https://github.com/elastic/elastic-agent/issues/3872.  Not sure if that changes the requirements for this PR or if a follow up PR will be needed to completely resolve that issue. Apologies if the original issue description was not clear enough.

**Comment by michalpristas** [comment]
the workflow i tested
- mage package agent
- sudo ./elastic-agent run -c otel.yml (alt test without `-c` to see it works) 
- keep it running
- other terminal `./elastic-agent status` it should show output


what are your expectation of running `./elastic-agent -c otel.yml status` on its own?

**Comment by ycombinator** [comment]
> what are your expectations of running `./elastic-agent -c otel.yml` status on its own?

None, I had a brain fade 😊.  Of course this cannot expected to work as we need a running Agent + GRPC server for `status` to talk to. 🤦 

Your workflow makes sense for running Agent in OTel mode.  I will test it with this PR.  Thanks.

**Comment by ycombinator** [comment]
Confirmed that the steps listed in https://github.com/elastic/elastic-agent/pull/4047#issuecomment-1888570168 work as expected. Reviewing the implementation now...

File: internal/pkg/agent/cmd/run.go [hunk 1]
```diff
@@ -136,17 +139,34 @@ func run(override cfgOverrider, testingMode bool, fleetInitTimeout time.Duration
...
+	if runAsOtel {
+		var otelStartWg sync.WaitGroup
+		otelAwaiter := make(chan struct{})
+		awaiters = append(awaiters, otelAwaiter)
+
+		otelStartWg.Add(1)
+		go func() {
+			otelStartWg.Done()
+			err := otel.Run(ctx, stop)
+			resErr = multierror.Append(resErr, err)
```
Comment (ycombinator) [inline internal/pkg/agent/cmd/run.go:154]: Maybe I missed something, but how is `resErr` being used?

File: internal/pkg/agent/cmd/run.go [hunk 2]
```diff
@@ -136,17 +139,34 @@ func run(override cfgOverrider, testingMode bool, fleetInitTimeout time.Duration
...
+
+			// close awaiter handled in run loop
+			close(otelAwaiter)
+		}()
+
+		// wait for otel to start
+		otelStartWg.Wait()
 	}
 
 	// not otel continue as usual
```
Comment (ycombinator) [inline internal/pkg/agent/cmd/run.go:164]: I think we should remove this comment, now that `otel.Run` could happen concurrently with `runElasticAgent` below?

**Review by ycombinator — APPROVED** [review]

**Comment by cmacknz** [comment]
I think this is probably mostly covered by integration tests and the SonarQube complaint can be bypassed.

> status reports only agent status (otel status not reported)

It seems to me like the integration test coverage could be improved to cover at least this, by running `elastic-agent status` in one of the otel tests and confirming it succeeds. Before this change this command would unconditionally fail wouldn't it?

Right now you have no net new test coverage at all, but there is actually added functionality to cover here.

**Comment by michalpristas** [comment]
status added to `TestOtelFileProcessing` test

**Review by cmacknz — APPROVED** [review]
Thanks for the test!

**Comment by cmacknz** [comment]
Force merging, ignoring the code coverage complaint because this is now covered in the integration tests.

**Pull request merged**